python/pyspark/sql/functions.py

```diff
@@ -2392,7 +2393,7 @@ def json_tuple(col, *fields):
 
 
 @since(2.1)
-def from_json(col, schema, options={}):
+def from_json(col, schema, options: Dict = None):
```

Comment:
I am happy with trying Python type hints. Actually, I am thinking about somehow leveraging (or porting?) https://github.com/zero323/pyspark-stubs, cc @zero323. Let's avoid to introduce the type hinting in the main source for now because we could decouple and run it separately.

Reply:
While touching the lines anyway, I thought is would be a nice addition. Having it in the source also makes live easier for the developers that work on pyspark itself. Curious what Maciej thinks.

Reply:
> Furthermore, it is easier to keep the code and annotations in sync. Curious what Maciej thinks.

That's the topic for a long discussion :) Overall, I made some points [here](http://apache-spark-developers-list.1001551.n3.nabble.com/Re-PySpark-Revisiting-PySpark-type-annotations-td26232.html) and [here](http://apache-spark-developers-list.1001551.n3.nabble.com/PYTHON-PySpark-typing-hints-td21560.html) ‒ I think these are still valid.

Now... I am still opened about merging `pyspark-stubs` about with the main project. However, have some annotations in the code-base, if there is no will to maintain these (and maintaining useful annotations is honestly a pain) and bring them to high coverage any time soon, as it may make things either wasted effort (if not used at all, if package is not PEP 561 compliant) or maybe even harmful. That's a topic for a longer discussion though...

In general I'd bring the discussion to dev and or JIRA first.

Reply:
Thanks for pointing out the maillist. Python <3.6 has been dropped in the meantime https://github.com/apache/spark/pull/28957.

I'm happy to help with migrating the annotations to the codebase. I still believe that it makes much sense to keep them sync with the actual code base. Otherwise you need to support multiple versions in your stubs and this also requires a lot of effort to keep them in sync.

Reply:
Pre 3.6 wasn't an issue anyway.. In practice, given the size of the project and overall complexity of annotations, stubs seem to be a somewhat better choice anyway (avoid cyclic imports or certain `Generic` pitfalls for starters, not being stuck in legacy Python support). This means that potential migration might be relatively easy.

My biggest concern is that over four years I haven't seen any active Spark contributors willing to help with maintenance (and that's despite gentle  pings). Additionally annotation ecosystem is still evolving ‒ so having a project that can iterate faster than annotated codebase is not a bad thing. 

The biggest advantage of maintaining annotations alongside actual codebase is that it requires conscious choices of signatures ‒ not every type of signature can be annotated in a meaningful way and some can result in exploding overloads. But it doesn't really remove the maintenance overhead ‒ for example changes in the ML API, mostly invisible for the final user, wreaked havoc, as it is the most vulnerable part of code (a lot of generics there). 

Anyway... That's just my 2 cents. If you reanimate the discussion. please ping me ‒ I'd like to see how it goes. TIA

Reply:
I'm happy to work on the annotations, but would be great to have some support from the committers as well. I'm planning to integrate mypy in the CI as well, so we can enforce the type checking in the main repository as well. Personally, I don't see any harm in adding the types over time, and work towards coverage, starting with the public functions because they add the most value.

Reply:
> Personally, I don't see any harm in adding the types over time, and work towards coverage,

The biggest problem is the resolution (https://bugs.python.org/issue39168 seems to be fixed now). As far as I am aware it is all-or-nothing at the module level (assuming `partial` mode), so working over time will likely degrade experience of the users that already use annotations.

And it seems somewhat unnecessary, considering that you could migrate stubs into main repo with a single copy, if there was a will to maintain the result. Inline variant would be trickier (and there is still concern of https://bugs.python.org/issue39168).

Reply:
How about continue the discussion [here](Given a discussion related to <a href="https://github.com/apache/spark/pull/29122">SPARK-32320 PR</a> I'd like to resurrect this thread. Is there any interest in migrating annotations to the main repository?) for a better visibility?

Reply:
The bug is fixed in 3.6, so that shouldn't be a problem. Better to move this to the maillist indeed, thanks!

Reply:
I've removed all the inline type annotations :)